File: baekjoon/dynamic/2169/main.cpp
#include <iostream>
#include <cstdio>
#include <cstring>

using namespace std;
const int oo = 0x3f3f3f3f;

int cache[1001][1001][2];
int map[1001][1001];
int n,m;
int solve(int row, int col, int prev);
int main()
{
    memset(cache, -1, sizeof(cache));
    scanf("%d %d", &n, &m);
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            scanf("%d", &map[i][j]);
        }
    }
    printf("%d\n", max(solve(0,0,0),solve(0,0,1)));
    return 0;
}

int solve(int row, int col, int d)
{
    if(row == n-1 && col == m-1) {
        return map[row][col];
    }
    
    int& ret = cache[row][col][d];
    if(ret != -1) return ret;
    ret = -oo;
    
    if(d == 0 && col-1>=0) {
        ret = max(ret, solve(row, col-1, d)+map[row][col]);
    }
    else if(d == 1 && col+1 <m) {
        ret = max(ret, solve(row, col+1, d)+map[row][col]);
    }
    
    if(row+1<n) {
        ret = max(ret, solve(row+1, col, 0)+map[row][col]);
        ret = max(ret, solve(row+1, col, 1)+map[row][col]);
    }
    return ret;
}

File: baekjoon/dynamic/2169/Main.java
import java.util.*;

class Main {
    static final int oo = 0x3f3f3f3f;
    static int [][] map;
    static int [][][] cache;
    static int n,m;
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        map = new int[n][m];
        cache = new int[n][m][2];
        
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                Arrays.fill(cache[i][j], -1);
                map[i][j] = sc.nextInt();
            }
        }
        
        System.out.println(max(solve(0,0,0), solve(0,0,1)));
    }
    
    public static int solve(int row, int col, int d) {
        if(row == n-1 && col == m-1) return map[row][col];
        
        if(cache[row][col][d] != -1) return cache[row][col][d];
        cache[row][col][d] = -oo;
        
        if(d == 0 && col-1 >= 0) {
            cache[row][col][d] = max(cache[row][col][d], solve(row,col-1,d)+map[row][col]);
        } else if(d == 1 && col+1 < m) {
            cache[row][col][d] = max(cache[row][col][d], solve(row,col+1,d)+map[row][col]);
        }
        
        if(row+1 < n) {
            cache[row][col][d] = max(cache[row][col][d], solve(row+1,col,0)+map[row][col]);
            cache[row][col][d] = max(cache[row][col][d], solve(row+1,col,1)+map[row][col]);
        }
        return cache[row][col][d];
    }
    
    public static int max(int a, int b) {
        return a>b? a : b;
    }
}